File: PracticeAndroid/AndroidPractice2/Task/RecyclerMVVM/app/src/main/java/com/example/recyclermvvm/MainActivity.java
package com.example.recyclermvvm;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;

import com.example.recyclermvvm.Model.NameModel;
import com.example.recyclermvvm.ViewModel.NameViewModel;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity implements DataLoadListener{

    private RecyclerView recyclerView;
    private  MyAdapter myAdapter;
    private NameViewModel nameViewModel;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        recyclerView=findViewById(R.id.recycler);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        nameViewModel= new ViewModelProvider(this).get(NameViewModel.class);
        nameViewModel.init(MainActivity.this);
        myAdapter=new MyAdapter(nameViewModel.getNames().getValue());
        recyclerView.setAdapter(myAdapter);
    }

    @Override
    public void OnNameLoaded() {
        nameViewModel.getNames().observe(this, new Observer<ArrayList<NameModel>>() {
            @Override
            public void onChanged(ArrayList<NameModel> nameModels) {
                myAdapter.notifyDataSetChanged();
            }
        });
    }
}

File: PracticeAndroid/Andorid_Kotlin/MVVMRetrofit/app/src/main/java/com/androdevdk/mvvmretrofit/MainActivity.kt
package com.androdevdk.mvvmretrofit

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.lifecycle.ViewModelProvider
import com.androdevdk.mvvmretrofit.api.QuoteService
import com.androdevdk.mvvmretrofit.api.RetrofitHelper
import com.androdevdk.mvvmretrofit.repository.QuoteRepository
import com.androdevdk.mvvmretrofit.viewmodel.MainViewModel
import com.androdevdk.mvvmretrofit.viewmodel.MainViewModelFactory
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {
    lateinit var mainViewModel: MainViewModel
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

     //   val quotesAPI = RetrofitHelper.getInstance().create(QuotesAPI::class.java)

       // val quoteService=RetrofitHelper.getInstance().create(QuoteService::class.java)
       // val repository=QuoteRepository(quoteService)
        val repository=(application as QuoteApplication).quoteRepository
        mainViewModel=ViewModelProvider(this,MainViewModelFactory(repository)).get(MainViewModel::class.java)

        mainViewModel.quotes.observe(this, {
           // Log.d("TAG", "onCreate: "+it.results.count().toString())
            Toast.makeText(this, it.results.count().toString(), Toast.LENGTH_LONG).show()

            val result=it.results
//            result.forEach {
//                Log.d("TAG", "onCreate: "+it.content.toString())
//            }
        })
//        GlobalScope.launch {
//            val result=quotesAPI.getQuotes(1)
//            if (result != null){
//               val quoteList = result.body()
//                if (quoteList != null) {
//                    quoteList.results.forEach{
//                        Log.d("TAG", "onCreate: "+it.content)
//                    }
//                }
//            }
//        }
    }
}